internal/upgrader/upgrader.go

```diff
@@ -47,37 +49,36 @@ func Upgrade(api CFClient, log Logger, cfg UpgradeConfig) error {
 		return performDeactivatedPlansCheck(api, cfg)
 	case config.CheckUpToDateAction:
 		return performUpToDateCheck(api, cfg)
-	default:
-		return performUpgradeOrDryRun(api, log, cfg)
+	default: // continue function
 	}
-}
 
-func performUpgradeOrDryRun(api CFClient, log Logger, cfg UpgradeConfig) error {
-	log.Printf("discovering service instances for broker: %s", cfg.BrokerName)
-	serviceInstances, err := getAllServiceInstances(api, cfg.BrokerName)
+	instances, err := getGroupedServiceInstances(api, cfg.BrokerName)
 	if err != nil {
 		return err
 	}
 
-	upgradableInstances := discoverInstancesWithPendingUpgrade(log, serviceInstances)
-
 	switch {
-	case len(upgradableInstances) == 0:
-		log.Printf("no instances available to upgrade")
-		return nil
-	case cfg.Action == config.DryRunAction:
-		log.InitialTotals(len(serviceInstances), len(upgradableInstances))
-		defer log.FinalTotals()
-		performDryRun(upgradableInstances, log)
-		return nil
+	case cfg.Action == config.DryRunAction && cfg.JSONOutput:
```

Comment:
Or maybe call a ```outputDryRun(upgradeable, failed, json)``` that encpasulates the output format option instead of two case options to make it more readable. But not feeling strongly. Happy with this.